### ISSUE TITLE
Deploy the model to ROS

### ISSUE THREAD

**Issue by Nuyoah0123** [engineerJPark/LiDARWeather#7]
Hello, I want to migrate this model to ROS to test its performance in a real environment, but I am not familiar with the training framework of mmengine, which makes me encounter some difficulties. 
My current idea is to receive the point cloud information of LiDAR through ROS, and then voxelize the point cloud through the `voxelize` function in `data_preprocessor.py`, and then get the predicted label of each point cloud through the forward propagation of the model. Finally, color the point cloud according to the label information and publish it. 
Unfortunately, I don't know how to initialize this model. I just extracted the code of the model described in `sj+lpd+minkunet_semantickitti.py`. If you can tell me how to initialize the model, I think I can complete this migration. 
I look forward to your reply.

**Comment by engineerJPark** [comment]
It was really long time ago that I used ROS before, so I'm not sure I can be helpful in this prob.
But I think there will be some way to do that without severe code change. Maybe ./tools/test.py file can be edit to be service node for ROS.

Check out this repo for reference. This may be helpful.
https://github.com/ksm26/ROS-based-3D-detection-Tracking

**Comment by Nuyoah0123** [comment]
> It was really long time ago that I used ROS before, so I'm not sure I can be helpful in this prob. But I think there will be some way to do that without severe code change. Maybe ./tools/test.py file can be edit to be service node for ROS.
> 
> Check out this repo for reference. This may be helpful. https://github.com/ksm26/ROS-based-3D-detection-Tracking

Thank you for your timely response. I instantiated the `runner class` using `Runner.from_cfg(cfg)` to initialize the model and load parameters, allowing me to input voxelized point clouds into the network for real-time segmentation. 

However, the results may not be optimal, likely due to differences in the LiDAR sensor used and its placement, along with the backbone network's performance potentially affecting the model's generalization. 

Nonetheless, your work remains highly valuable, and I plan to delve into your paper further; I hope we can exchange ideas again in the future.

**Comment by engineerJPark** [comment]
I wonder you're doing LiDAR segmentation in real world, right?
I also recently learned from another paper that placement affects the domain gap. I feel that to achieve generalization across all domains—not just weather—and to attain true real-world domain generalization, there are many additional tasks that need to be studied. But those are out of the scope in this work...

As far as I remember, Lingdong Kong's research involves quite a number of these tasks. If you are also interested in this area of research, it would be great if you could read his work (or other works) and continue exchanging opinions in this area.
Thank you for your question and attention to my work.

**Comment by Nuyoah0123** [comment]
> I wonder you're doing LiDAR segmentation in real world, right? I also recently learned from another paper that placement affects the domain gap. I feel that to achieve generalization across all domains—not just weather—and to attain true real-world domain generalization, there are many additional tasks that need to be studied. But those are out of the scope in this work...
> 
> As far as I remember, Lingdong Kong's research involves quite a number of these tasks. If you are also interested in this area of research, it would be great if you could read his work (or other works) and continue exchanging opinions in this area. Thank you for your question and attention to my work.

Yes, I tested the performance of the model on my own experimental car. I also studied the related work of Lingdong Kong. I think generalization of Lidar semantic segmentation domain is a very challenging task, and I will continue to explore this field. 🤝
